### DISCUSSION TITLE
Displaced response when send two different request

### DISCUSSION THREAD

**Discussion by javadghasemi** [miguelgrinberg/python-socketio#722] [Q&A]
I have a nodejs server and I want to connect to my python client. When the client sends a request to a specific URL, nodejs sends data to the python client and python sends a response back to nodejs. I use socketio for nodejs server and python socketio for python client. Unfortunately, the python-socketio client doesn't support rooms. Then when I send two requests with different data, responses are displaced. I'm trying to find a way to solve this problem. Any help would be appreciated.

**Comment by miguelgrinberg** [comment]
> the python-socketio client doesn't support rooms

What do you mean here? Does the JavaScript Socket.IO client support rooms? This is a server feature, not a client feature.

**Comment by javadghasemi** [comment]
jisses!! I have just one client connected to my server. In my client, I have many functions and my functions have a timeout c a response. When I send two requests to two different functions, returning values are displaced.
how can I fix this?? 
I thinged the client has room and I can make room for any of the functions.
But you said it's just a server feature.

Please give me a solution.

**Comment by miguelgrinberg** [comment]
I'm sorry, but you need to make an effort to explain the situation better.

You said:

- In my client, I have many functions
- my functions have a timeout
- When I send two requests to two different functions

This does not make sense. You would be sending requests to the server, not to the client. What do you mean by functions here? Why don't you create a small example application that shows the problem? That will help me understand your problem.

**Comment by javadghasemi** [comment]
Client:
```

async def main():
    # Sockets config
    sio = socketio.AsyncClient()
    await sio.connect("http://localhost:8082")

    # Socket on connection event
    @sio.on('connect')
    def connect():
        print("Python connected successfully")

    # Socket on connection Error event
    @sio.event
    def connect_error():
        print("The connection failed!")

    # socket on Disconnect event
    @sio.event
    def disconnect():
        print("I'm disconnected!")


    # Socket on Run event
    @sio.on('run')
    async def run_function(data):
        return await sio.emit('response', data['functionName'](data['data']))




    def sample_function1():
        return "file: sample | This is sample_function1 from sample file"


    def sample_function2():
        return "file: sample | This is sample_function2 from sample file"


    def test_starter(arg):
        if arg['body']['step'] == 'first':
            return 'im first'
        elif arg['body']['step'] == 'second':
            time.sleep(5)
            return 'im second'
        elif arg['body']['step'] == 'third':
            time.sleep(10)
            return 'im third'

    await sio.wait()


loop = asyncio.get_event_loop()
loop.run_until_complete(main())
loop.close()


```

Server:

```
const sockets = require('socket.io')(8082)

function runScript(functionName, query, body) {
    return new Promise(((resolve, reject) => {

      sockets.to('python').emit('run', {functionName, data: {query, body}});
      
      sockets.on('connection', function (socket) {
         const handleResponse = (res) => {
            if (res) {
              resolve({'message': res, 'status': 'ok'});
            }
            else {
              reject({'message': 'Function is not found', 'status': 'failed'});
            }
         };
  
         socket.on('response', handleResponse);

         socket.on('disconnect', function () {
           console.log("disconnected")
         });
      });
    }));
  }
```

When I start `test_starter` simultaneously with different args, the response displaced.

example:
```
runScript("test_runner", {}, {"step": "second"}).then(result => ...); // result: second
runScript("test_runner", {}, {"step": "third"}).then(result => ...);  // result: second
```

**Comment by miguelgrinberg** [comment]
This isn't a complete example. There is nothing that calls your `test_starter` function. Also this function has nothing related to Socket.IO, so I'm still at a loss regarding the problem you are seeing.
